File: app/src/main/java/com/aquaguard/presentation/alerts/AlertsViewModel.java
package com.aquaguard.presentation.alerts;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.aquaguard.domain.model.Alert;
import com.aquaguard.domain.model.Device;
import com.aquaguard.domain.repository.AlertRepository;
import com.aquaguard.domain.repository.DeviceRepository;
import com.aquaguard.domain.repository.RepositoryCallback;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class AlertsViewModel extends ViewModel {
    private final DeviceRepository deviceRepository;
    private final AlertRepository alertRepository;

    private final LiveData<List<Device>> devices;
    private final MutableLiveData<String> selectedDeviceId = new MutableLiveData<>(null);
    private final MediatorLiveData<List<Alert>> activeAlerts = new MediatorLiveData<>();
    private final MediatorLiveData<List<Alert>> alertHistory = new MediatorLiveData<>();

    private LiveData<List<Alert>> currentActiveAlertsSource = null;
    private LiveData<List<Alert>> currentAlertHistorySource = null;

    @Inject
    public AlertsViewModel(DeviceRepository deviceRepository, AlertRepository alertRepository) {
        this.deviceRepository = deviceRepository;
        this.alertRepository = alertRepository;

        this.devices = deviceRepository.getDevices();

        // Select first device by default
        this.activeAlerts.addSource(this.devices, list -> {
            if (selectedDeviceId.getValue() == null && list != null && !list.isEmpty()) {
                selectDevice(list.get(0).getDeviceId());
            }
        });

        // Switch source on device change
        this.activeAlerts.addSource(selectedDeviceId, id -> {
            if (currentActiveAlertsSource != null) {
                activeAlerts.removeSource(currentActiveAlertsSource);
            }
            if (id != null) {
                currentActiveAlertsSource = alertRepository.getActiveAlerts(id);
                activeAlerts.addSource(currentActiveAlertsSource, activeAlerts::setValue);
            } else {
                currentActiveAlertsSource = null;
                activeAlerts.setValue(new ArrayList<>());
            }
        });

        this.alertHistory.addSource(selectedDeviceId, id -> {
            if (currentAlertHistorySource != null) {
                alertHistory.removeSource(currentAlertHistorySource);
            }
            if (id != null) {
                currentAlertHistorySource = alertRepository.getAlertHistory(id);
                alertHistory.addSource(currentAlertHistorySource, alertHistory::setValue);
            } else {
                currentAlertHistorySource = null;
                alertHistory.setValue(new ArrayList<>());
            }
        });
    }

    public LiveData<List<Device>> getDevices() {
        return devices;
    }

    public LiveData<String> getSelectedDeviceId() {
        return selectedDeviceId;
    }

    public LiveData<List<Alert>> getActiveAlerts() {
        return activeAlerts;
    }

    public LiveData<List<Alert>> getAlertHistory() {
        return alertHistory;
    }

    public void selectDevice(String deviceId) {
        selectedDeviceId.setValue(deviceId);
    }

    public void resolveAlert(String alertId) {
        alertRepository.resolveAlert(alertId, new RepositoryCallback<Void>() {
            @Override
            public void onSuccess(Void result) {
                // Done
            }

            @Override
            public void onFailure(Exception e) {
                // Handle failure
            }
        });
    }
}

File: Aqua Guard/frontend/app/src/main/java/com/aquaguard/presentation/alerts/AlertsViewModel.kt
package com.aquaguard.presentation.alerts

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.aquaguard.domain.model.Alert
import com.aquaguard.domain.model.Device
import com.aquaguard.domain.repository.AlertRepository
import com.aquaguard.domain.repository.DeviceRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

@OptIn(ExperimentalCoroutinesApi::class)
@HiltViewModel
class AlertsViewModel @Inject constructor(
    private val deviceRepository: DeviceRepository,
    private val alertRepository: AlertRepository
) : ViewModel() {

    val devices: StateFlow<List<Device>> = deviceRepository.getDevices()
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    private val _selectedDeviceId = MutableStateFlow<String?>(null)
    val selectedDeviceId = _selectedDeviceId.asStateFlow()

    val activeAlerts: StateFlow<List<Alert>> = _selectedDeviceId
        .flatMapLatest { id ->
            if (id != null) alertRepository.getActiveAlerts(id) else flowOf(emptyList())
        }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    val alertHistory: StateFlow<List<Alert>> = _selectedDeviceId
        .flatMapLatest { id ->
            if (id != null) alertRepository.getAlertHistory(id) else flowOf(emptyList())
        }
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5000), emptyList())

    init {
        viewModelScope.launch {
            devices.collect { list ->
                if (_selectedDeviceId.value == null && list.isNotEmpty()) {
                    _selectedDeviceId.value = list.first().deviceId
                }
            }
        }
    }

    fun selectDevice(deviceId: String) {
        _selectedDeviceId.value = deviceId
    }

    fun resolveAlert(alertId: String) {
        viewModelScope.launch {
            alertRepository.resolveAlert(alertId)
        }
    }
}
